designs/mixins.md

```diff
@@ -0,0 +1,1075 @@
+```
+structure ListCitiesInput {
+    nextToken: String,
+    pageSize: Integer
+}
+
+structure ListCitiesOutput {
+    nextToken: String,
+
+    @required
+    items: CitySummaries,
+}
+
+structure ListWeatherStationsInput {
+    nextToken: String,
+    pageSize: Integer
+}
+
+structure ListWeatherStationsOutput {
+    nextToken: String,
+
+    @required
+    items: WeatherStationSummaries,
+}
+```
+
+This repetition is tedious and error-prone as it can lead to unintentional
+inconsistencies due to copy/paste errors and drift between related shapes
+over time.
+
+
+## Proposal
+
+This proposal introduces *mixins* to reduce the amount of repetition in
+structures and unions, reduce copy/paste errors, and provide the ability to
+define reusable partial structures and unions.
+
+
+### Goals
+
+1. Improve maintainability of large models by reducing repetition and
+   copy/paste errors of related concepts.
+2. Mixins are an implementation detail of the model that have no impact on
+   generated code or collaborators that need to apply traits to shapes that use
+   mixins.
+3. Allow refinement of mixin traits so that a mixin can provide reasonable
+   default traits, and shapes that use mixins can override them with more
+   specific traits.
+4. Reduce the surface area for bugs by treating mixin members that are copied
+   onto structures just like normal members when using the Smithy meta model.
+
+
+### Overview
+
+A mixin is a structure or union marked with the `@mixin` trait,
+`smithy.api#mixin`:
+
+```
+$version: "1.1"
+
+namespace smithy.example
+
+@mixin
+structure CityResourceInput {
+    @httpLabel
+    @required
+    cityId: String
+}
+```
+
+The `with` clause of a structure or union statement is used to merge the members
+of one or more mixins into a structure or union. Each shape ID in the `with`
+clause MUST target a shape marked with the `@mixin` trait. Structure shapes can
+only use structure mixins, and union shapes can only use union mixins.
+
+```
+structure GetCityInput with CityResourceInput {}
+```
+
+Multiple mixins can be applied:
+
+```
+structure GetCityInput with CityResourceInput, SomeOtherMixin {}
+```
+
+Mixins can be composed of other mixins:
+
+```
+@mixin
+structure A {
+    a: String
+}
+
+@mixin
+structure B with A {
+    b: String
+}
+
+structure C with B {
+    c: String
+}
+```
+
+When a member is copied from a mixin into a target shape, the shape ID of the
+copied member takes on the containing shape ID of the target shape. This
+ensures that members defined via mixins are treated the same way as mixins
+defined directly in a shape, and it allows members of a shape to be backward
+compatibly refactored and moved into a mixin or for a shape to remove a mixin
+and replace it with members defined directly in the shape.
+
+The above `C` structure is equivalent to the following flattened structure:
+
+```
+structure C {
+    a: String,
+    b: String,
+    c: String
+}
+```
+
+Unions can be mixins and use mixins too.
```

Comment:
Can union mixins be used with structures and vice versa? e.g.

```
@mixin 
union UnionMixin {
    a: String
}

structure Struct with UnionMixin {
    b: String
}
```

Pretty sure the answer is no, but it should be said explicitly.

Reply:
No. There's a callout for this that states:
> Structure shapes can only use structure mixins, and union shapes can only use union mixins.